torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -320,48 +332,67 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
+        shard_dim = fsdp_placement.dim
+        if shard_dim >= param_data.ndim:
+            raise AssertionError(
+                f"Shard dim {shard_dim} is invalid for {param_data.ndim}D tensor: {param.shape}"
+            )
         self._orig_size = param_data.size()
         self._contiguous_orig_stride = make_contiguous_strides_for(self._orig_size)
         shard_rank = self.mesh_info.shard_mesh_rank
         shard_world_size = self.mesh_info.shard_mesh_size
-        chunks = _chunk_with_empty(param_data, shard_world_size, dim=0)
+        if shard_dim > 0 and param_data.size(shard_dim) % shard_world_size != 0:
+            # If sharding on nonzero dim, require even sharding for now because
+            # the uneven sharding (1) requires extra copies before/after FSDP
+            # collectives and (2) introduces extra complexity to handle padding
+            # and unpadding
+            raise NotImplementedError(
+                f"FSDP does not support uneven sharding on dim {shard_dim}: {param_data.size()}"
+            )
+        chunks = _chunk_with_empty(param_data, shard_world_size, dim=shard_dim)
         sharded_param = chunks[shard_rank]
-        self.sharded_size = _get_dim0_chunked_size(sharded_param, param_data.size())
+        self.sharded_size = _get_dim_chunked_size(
+            sharded_param, param_data.size(), dim=shard_dim
+        )
         self.contiguous_sharded_stride = make_contiguous_strides_for(self.sharded_size)
         padded_sharded_size = chunks[0].size()  # 0th always padded
+        self.padded_sharded_param_size = padded_sharded_size
+        # Pre-pad the sharded parameter to avoid padding before all-gather
         padded_sharded_param = param_data.new_zeros(padded_sharded_size)
-        self.padded_sharded_param_size = padded_sharded_param.size()
         if sharded_param.numel() > 0:
-            padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
+            padded_sharded_param.narrow(
+                dim=shard_dim, start=0, length=sharded_param.size(shard_dim)
+            ).copy_(sharded_param)
         if self.offload_to_cpu and not padded_sharded_param.is_meta:
             padded_sharded_param = padded_sharded_param.cpu()
             if self.pin_memory:
                 padded_sharded_param = padded_sharded_param.pin_memory(
                     device=self.device
                 )
         self._sharded_param_data = padded_sharded_param.view(-1)
-        self.sharded_param = nn.Parameter(
-            self.to_sharded_dtensor(padded_sharded_param[: sharded_param.size(0)])
-        )
+        if padded_sharded_param.numel() > 0:
+            length = sharded_param.size(shard_dim) if sharded_param.numel() > 0 else 0
+            sharded_param = padded_sharded_param.narrow(
+                dim=shard_dim, start=0, length=length
+            )
+        else:
+            sharded_param = padded_sharded_param.view(self.sharded_size)
```

Comment:
curious when will we trigger this branch, `padded_sharded_param.numel() == 0`? is this to cover the edge case when `model.parameters()` returns a param with shape `()` ?

Reply:
This is when the rank only has padding for the sharded parameter. This code was useful when we supported uneven sharding for `Shard(i)` so that the shape is like `[256, 0]` instead of just `[0]` for example. It might not be needed anymore.